### PR TITLE
Re-do rent collection check on rent-exempt account

### LINKED ISSUE #11342: Possible resource exhausition attack for account storage with reasonable funds
#### Problem

Currently, the cluster doesn't further collect rents for the duration of current epoch if previous rent has been already collected for the epoch.
A determined attacker can exploit this behavior of rather lenient rent check and can mount a temporal resource exhaustion attack within an epoch with reasonable fund like this:

1. Create `MAX_PERMITTED_DATA_LENGTH`-sized account with the balance of `rent_exempt_for(MAX_PERMITTED_DATA_LENGTH)` with 5,000 transaction fee.
1. Transfer all of the balance out of the newly-created account but 1 lamports (which leaves a huge account rent-exempt for the epoch) with 5,000 transaction fee.
1. Repeat the above

Each iteration effectively costs 10001 lamports and only takes minimum of 2 slots while that would forces the cluster to hold 10M bytes for duration of an epoch.
Also, this can easily parallelized.

So, attacker can consume 500TB with 500 concurrent processing with about 550 sols and ~24 hours, causing the trouble for the remainder of the epoch (~1 day):

```
[11] pry(main)> ((500 * 1024 * 1024 * 1024 * 1024) / (10 * 1024 * 1024)) * 10001
=> 524340428800
[14] pry(main)> (((500 * 1024 * 1024 * 1024 * 1024) / (10 * 1024 * 1024)) / 500) * 2 / 2.5 / 3600
=> 23.301555555555556
```

Worse yet, if this attack is done persistently, the resource exhaustion attack can continue until the attack ends + 1 epoch. Upper bound to 1PB resource exhaustion with 1100 sols per epoch for sustained cost with the proceeding assumption.

Alternatively, this could be bound to the IO bandwidth to the `mmap()`-ed account storage...

Anyway, this is well above the worst-case resource assumption across our current validators.

#### Proposed Solution

Adjust to increase rent check timing?

#### Context

Found via thinking for the fix of #10468

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

Once passed as a rent-exempt for this epoch, accounts can reside with a penny (= 1 lamport) for the duration of the epoch. This could be exploited to occupy huge amount of storage for accounts, leading to a DOS.

#### Summary of Changes

rent-collect again only on previously-rent-exempt accounts.

~Alternatively, we can use the highest bit of the `account.rent_epoch`? or introduce a new bit field? This would introduce abi breakage, though...~ => Well, it turns out the fix is pretty simple.

With this fix in effect, attackers actually must leave enough lamports for 10 * 1024 * 1024-byte MAX_PERMITTED_DATA_LENGTH accounts (= roughly 0.2 SOL (a)) to avoid it from being reclaimed after becoming non-rent-exempt.

So, it costs 10M SOL to do 1-epoch 500TB account storage DOS (b):

```
[1] pry(main)> 19.055441478439427 * 10 * 1024 * 1024 # <- (a)
=> 199810786.03696102
[5] pry(main)> (500 * 1024 * 1024 * 1024 * 1024) / (10 * 1024 * 1024) * (199810786.03696102 / 1000000000) # <- (b)
=> 10475839.738974622
```

, meaning we're safe now.

Fixes #11342

### CODE REVIEW

File: runtime/src/rent_collector.rs [hunk 1]
```diff
@@ -33,46 +33,66 @@ impl RentCollector {
...
         if account.executable
-            || account.rent_epoch > self.epoch
             || sysvar::check_id(&account.owner)
             || *address == incinerator::id()
         {
             0
+        } else if account.rent_epoch == self.epoch + 2
+            // previously this account was rent exempt
+            // this depends on no bug in the eager rent collection
+            // also needs gating
```
Comment (ryoqun) [inline runtime/src/rent_collector.rs:49]: ... so this pr is blocked on #10810 and #10809

File: runtime/src/rent_collector.rs [hunk 2]
```diff
@@ -33,46 +33,66 @@ impl RentCollector {
...
-                if account.lamports > rent_due {
-                    account.rent_epoch = self.epoch + 1;
-                    account.lamports -= rent_due;
-                    rent_due
+    fn update_account(account: &mut Account, rent_due: u64, exempt: bool) -> u64 {
+        if exempt || rent_due != 0 {
+            if account.lamports > rent_due {
+                account.rent_epoch = self.epoch + if exempt {
+                    // this way, mark last rent collection was exempt...
+                    2
```
Comment (ryoqun) [inline runtime/src/rent_collector.rs:82]: wait we just could use +0, and everything works nicely???

File: runtime/src/rent_collector.rs [hunk 3]
```diff
@@ -33,46 +33,66 @@ impl RentCollector {
...
     //  the account rent collected, if any
     //
     pub fn update(&self, address: &Pubkey, account: &mut Account) -> u64 {
         if account.executable
-            || account.rent_epoch > self.epoch
             || sysvar::check_id(&account.owner)
             || *address == incinerator::id()
         {
             0
+        } else if account.rent_epoch == self.epoch + 2
```
Comment (ryoqun) [inline runtime/src/rent_collector.rs:46]: also just make this `rent_epoch == self.epoch`?

**Comment by ryoqun** [comment]
@t-nelson @mvines Could one of you sanity check this? Once granted, I'll do the remaining work (test fixes, gating).

Also, I guess we should fix this with reasonably quick fashion. :)

**Comment by ryoqun** [comment]
todo
- [ ] rebase on #10468 
- [ ] doc update for rent_epoch + 1 invariant no longer valid and fyi

**Comment by ryoqun** [comment]
@t-nelson Thanks for early-review. I think this pr is now ready for serious review to merge. :)

File: docs/src/implemented-proposals/rent.md [hunk 4]
```diff
@@ -24,11 +24,11 @@ Even if those processes are out of scope of rent collection, all of manipulated
 
 ## Actual processing of collecting rent
 
-Rent is due for one epoch's worth of time, and accounts always have `Account::rent_epoch` of `current_epoch + 1`.
+Rent is due for one epoch's worth of time, and accounts have `Account::rent_epoch` of `current_epoch` or `current_epoch + 1` depending on the rent regime.
```
Comment (ryoqun) [inline docs/src/implemented-proposals/rent.md:27]: @rwalker-apple Hi, I'm yet again tweaking the rent protocol a bit, mainly affected by this dos possibility: #10847. Could you review this in your free time?

File: core/src/rpc_pubsub.rs [hunk 5]
```diff
@@ -582,7 +582,7 @@ mod tests {
                        "lamports": 51,
                        "data": bs58::encode(expected_data).into_string(),
                        "executable": false,
-                       "rentEpoch": 1,
+                       "rentEpoch": 0,
```
Comment (ryoqun) [inline core/src/rpc_pubsub.rs:585]: @CriesofCarrots @jstarry Strictly speaking, we're breaking some semantics (=~ compatibility) of our rpc api a bit here according to this new rule for `rentEpoch`: https://github.com/solana-labs/solana/pull/11349/files#r469691300.

It's an incompatible change, but the change is soooo subtle, I think. So, I believe there should be virtually no one relying on `rentEpoch` being the current epoch or the next in any-conceivable way for the bizarre usecases.

Comment (ryoqun) [reply core/src/rpc_pubsub.rs:585]: Or, if you have something to come to mind in terms of various partner integartions, please let me know. :)

File: runtime/src/rent_collector.rs [hunk 6]
```diff
@@ -74,7 +98,15 @@ impl RentCollector {
...
-                    account.rent_epoch = self.epoch + 1;
+                    account.rent_epoch = self.epoch
+                        + if !self.enable_old_behavior && exempt {
+                            // Rent isn't collected for the next epoch
+                            // Make sure to check exempt status later in curent epoch again
+                            0
+                        } else {
+                            // Rent is collected for next epoch
+                            1
+                        };
```
Comment (ryoqun) [inline runtime/src/rent_collector.rs:109]: This is the crux of this pr.

File: core/src/rpc_pubsub.rs (hunk 5 above)
Comment (jstarry) [reply core/src/rpc_pubsub.rs:585]: @ryoqun agreed, very unlikely this will cause any issues. I highly doubt anyone is using `rentEpoch` right now

Comment (CriesofCarrots) [reply core/src/rpc_pubsub.rs:585]: I don't know of any integrations using `rentEpoch`; I'm comfortable with this breaking change.

File: runtime/src/rent_collector.rs [hunk 7]
```diff
@@ -30,21 +40,35 @@ impl RentCollector {
...
     }
+
+    fn enable_old_behavior(epoch: Epoch, operating_mode: OperatingMode) -> bool {
+        let enable_new_behavior = match operating_mode {
+            OperatingMode::Development => true,
+            OperatingMode::Preview => epoch >= Epoch::max_value(),
+            OperatingMode::Stable => epoch >= Epoch::max_value(),
+        };
+
+        !enable_new_behavior
```
Comment (t-nelson) [inline runtime/src/rent_collector.rs:69]: NIT: The `!`ing back and forth here and in the next hunk is a little clumsy to follow.  Maybe just settle on one? (new seems like it'd read best across the board)

File: programs/bpf/c/src/invoked/invoked.c [hunk 8]
```diff
@@ -36,7 +36,7 @@ extern uint64_t entrypoint(const uint8_t *input) {
     sol_assert(accounts[ARGUMENT_INDEX].data_len == 100);
     sol_assert(accounts[ARGUMENT_INDEX].is_signer);
     sol_assert(accounts[ARGUMENT_INDEX].is_writable);
-    sol_assert(accounts[ARGUMENT_INDEX].rent_epoch == 1);
+    sol_assert(accounts[ARGUMENT_INDEX].rent_epoch == 0);
```
Comment (ryoqun) [inline programs/bpf/c/src/invoked/invoked.c:39]: @jackcmay Like [the RPC land](https://github.com/solana-labs/solana/pull/11349/files#r469692568), this pr changes slightly changes [the semantics of `rent_epoch`](https://github.com/solana-labs/solana/pull/11349/files#r469691300), which is exposed to programs as well. Are you aware of any programs strictly relying on the current semantics of `rent_epoch`?

**Comment by jackcmay** [comment]
Not that I know of

File: runtime/src/rent_collector.rs (hunk 7 above)
Comment (ryoqun) [reply runtime/src/rent_collector.rs]: yeah, I've been bothered with some convoluted handling for gating... Now your mentioning this encouraged me to do the correct thing: https://github.com/solana-labs/solana/pull/11349/commits/047b43ab115d92937812e79ded7e0d3dddf0a58b. :)

**Comment by ryoqun** [comment]
@t-nelson Thanks for re-approving! Now waiting for last-minute local testing by me and final say from @rwalker-com / @rwalker-apple (if any).

**Comment by rwalker-com** [comment]
lgtm

**Comment by ryoqun** [comment]
> Now waiting for last-minute local testing by me

This is done finally and merged this!

**Pull request merged**